### ISSUE TITLE
Switch to langchain

### ISSUE THREAD

**Issue by Yiannis128** [Yiannis128/esbmc-ai#53]
Consider switching to [langchain](https://github.com/hwchase17/langchain). Need to consider the positives and negatives.

**Comment by krrishdholakia** [comment]
Hey @Yiannis128 cool repo - curious, why switch to Langchain? was it to support Falcon?

**Comment by Yiannis128** [comment]
> Hey @Yiannis128 cool repo - curious, why switch to Langchain? was it to support Falcon?

Hi, short answer: yes. Slightly longer answer is that it will also allow for easily adding other types of LLMs such as Google PaLM that all have different APIs. Langchain makes the issue of designing, implementing, and testing a custom interface go away.

The transition is mostly complete, with only a slight set of features left to achieve feature parity with pre-langchain ESBMC-AI.

**Comment by krrishdholakia** [comment]
Curious - would this have helped - https://github.com/BerriAI/litellm?

**Comment by Yiannis128** [comment]
> Curious - would this have helped - https://github.com/BerriAI/litellm?

Nice suggestion, I will have to check it out. The current implementation of LangChain is very generic (only uses API calls for completions), so, in theory, I could replace the LangChain support with LiteLLM.

I would have to see if:
* LiteLLM is developed and supported regularly—wouldn't want to switch to something that is not going to be updated in the future.
* Does it contain a good API with features, LangChain is OK for this, so switching would only occur if the features offered are better.
* Is it worth it? If it's more lightweight, then maybe...

**Comment by krrishdholakia** [comment]
Any specific features you're looking for / problem you're facing with current implementation? Happy to submit a PR to help out here 😊

**Comment by Yiannis128** [comment]
Hello, sorry for the late reply. I am looking for easy interop between the different APIs

**Comment by krrishdholakia** [comment]
Hey @Yiannis128, 

No worries. I went through the code and here's what I understand:
- You currently support OpenAI + Huggingface Text Gen API (falcon + starcoder). 
-  For Huggingface Text Gen API you have some logic for translating I/O before/after making the call 

How do you expect users to make the call to falcon-7b or starcoder? would they have to deploy it themselves via huggingface inference api before using esbmc? both falcon-7b and starcoder are available pretty easily on other providers - e..g Baseten which also offers free credit

**Comment by krrishdholakia** [comment]
<img width="1580" alt="Screenshot 2023-08-26 at 6 20 44 AM" src="https://github.com/Yiannis128/esbmc-ai/assets/17561003/dc541a5e-4ce5-4de2-9240-d829735fc60c">

**Comment by Yiannis128** [comment]
> Hey @Yiannis128, 
> 
> No worries. I went through the code and here's what I understand:
> - You currently support OpenAI + Huggingface Text Gen API (falcon + starcoder). 
> -  For Huggingface Text Gen API you have some logic for translating I/O before/after making the call 

Yeah that's right, as the hugging face API is more generic due to the diverse amount of models it supports.

> How do you expect users to make the call to falcon-7b or starcoder? would they have to deploy it themselves via huggingface inference api before using esbmc? both falcon-7b and starcoder are available pretty easily (1-click deploy) on other providers - e..g Baseten which also offers free credit

No need to use one click deploy for some models. The ones built-in to ESBMC-AI make calls to hugging face servers, as they're hosted free of charge.

Larger models need to be added through the config as custom AI models that are hosted elsewhere.
